### ISSUE TITLE
Efficient leave-one-out cross-validation?

### ISSUE THREAD

**Issue by Tal-Golan** [nrdg/fracridge#38]
Is there an easy way of combing fractional RR with efficient leave-one-out cross-validation?

SKLearn has a documented implementation of LOOCV for standard RR:
https://github.com/scikit-learn/scikit-learn/blob/7e1e6d09bcc2eaeba98f7e737aac2ac782f0e5f1/sklearn/linear_model/_ridge.py#L1432

**Comment by kendrickkay** [comment]
I checked out the links you sent; thanks.  I haven't fully wrapped my head around the logic for the LOOCV idea in the technical reports.  It sounds like perhaps that if you get a range of ridge-regression solutions (e.g. fractions from 0 to 1 in steps of 0.01 or 0.02 or something) for a full dataset, there might be some analytic solutions for estimating/determining what the leave-one-out error would have been for each of the ridge-regression solutions?  If so, then that could greatly speed up computational time with respect to actually doing the leave-one-out procedure.

**Comment by Tal-Golan** [comment]
Yes, that's exactly my understanding. This is explained in more detail here:
https://www.mit.edu/~9.520/spring07/Classes/rlsslides.pdf
(slides 14-20, slide 30 for the SVD form)

**Comment by kendrickkay** [comment]
Yeah...  but I have trouble parsing it.  Have you figured it out / coded it with a simple example maybe?

**Comment by Tal-Golan** [comment]
https://gist.github.com/Tal-Golan/47e0daecf77d800136125576417a7450

**Comment by kendrickkay** [comment]
Nice! Do you want to Zoom to discuss a bit? I don't quite speak Python. Also, it seems too good to be true. Is it verifiable that the solution is exactly equal to what would happen if the full LOO procedure was implemented? Are there any limitations/assumptions going on?  Maybe email me (kendrickkay@gmail.com) to find a time to Zoom?

**Comment by Tal-Golan** [comment]
Sure, I'll email you. We can empirically verify the equivalence of the results with algorithmic LOOCV for different regimes (e.g., not full rank, more features than observations). I don't see explicit assumptions in Rifkin's notes.

**Comment by kendrickkay** [comment]
I did some simulations to verify the correctness. It does seem promising.

<img width="477" alt="image" src="https://user-images.githubusercontent.com/2728338/153687460-8c5bf2a8-bd47-4494-91c0-c44e632aa57d.png">
However, I noticed something. It seems that an implementation would, in theory, have to do a for-loop over each alpha value and for each target (e.g. see screenshot).   This is because, technically, to achieve a nice uniform set of fractions for every target, the corresponding alpha that would achieve those fractions might in fact vary from target to target.  (Of course, in practice, the variation may be small...)    If that is true, then the numerically correct implementation would, ironically, be really slow in the case of many targets.   That seems not very appealing.  So, then, if this is all true, then ironically a "n-fold" cross-validation scheme that the user performs based on the returned coefficients by fracridge might be faster.

**Comment by Tal-Golan** [comment]
I agree that such a nested loop over targets and alphas isn't ideal, but this might still be worthwhile for usage cases in which the number of targets is not huge. Maybe the simplest solution is to add an optional call to a function within the target loop (around line 200 in fracridge.py) that will estimate LOOCV MSE per alpha and target? this targets X alphas MSE matrix can be then stored as a FracRidgeRegressor object property.

**Comment by kendrickkay** [comment]
good idea.   I've been playing around with the matlab implementation.  
![image](https://user-images.githubusercontent.com/2728338/153784643-294f068e-fbcc-4d49-8373-50cfdf25b8ab.png)
this is sort of cool. x-axis is log(alpha). y-axis is MSE. red is a naive LOO implementation. blue is the analytic method. black is the fracridge analytic LOO method.  Notice how black follows the red and blue, but that black is more sensibly arranged on the x-axis (which is the whole point of fracridge). Notice also that fracridge finds a (slightly) better minimum.

**Comment by Tal-Golan** [comment]
Yes, this is indeed a convincing demonstration of using fracridge. One more argument in favor of including the analytic LOO is that it would reduce the common mistake of using the test set to choose the alpha.

**Comment by kendrickkay** [comment]
<img width="764" alt="image" src="https://user-images.githubusercontent.com/2728338/153790002-63b01966-0efd-46c6-9aed-1381cbe791d3.png">

Other examples are more extreme, see above.

Yeah, I suppose that bad users can do bad things

**Comment by kendrickkay** [comment]
I just pushed an update to the MATLAB fracridge.m with the new feature.

here is an example (also in the documentation of the function):
![image](https://user-images.githubusercontent.com/2728338/153790239-0ce740d2-54f2-4c6f-8c8b-41a5aa3c8c51.png)

**Comment by kendrickkay** [comment]
If you speak MATLAB and want to try it, I would be curious.

Otherwise, we will have to wait on @arokem if he can port the update to Python.

**Comment by kendrickkay** [comment]
(Passing the nerd sniping to @arokem )

**Comment by Tal-Golan** [comment]
better tools make better users... I speak Matlab but need to renew my license. I'll test the Python code as soon as @arokem is done.
